File: Backtracking/generateSubsets.java
import java.util.*;
class generateSubsets{
    public static void main(String[] args) {
        int[] arr = { 1, 2, 3 };
        List<List<Integer>> res = generate(arr);
        System.out.println(res);
    }

    public static List<List<Integer>> generate(int[] arr) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(arr);
        backtrack(res, new ArrayList<>(), arr, 0);
        return res;
    }

    private static void backtrack(List<List<Integer>> res,
            List<Integer> curSubset,int[] arr, int start) {
        res.add(new ArrayList<>(curSubset));
        for (int i = start; i < arr.length; i++) {
            curSubset.add(arr[i]);
            backtrack(res, curSubset, arr, i + 1);
            curSubset.remove(curSubset.size() - 1); // backtrack
        }
    }
}

File: Backtracking/generateSubsets.py
arr=[1,2,3]

def generateSubsets(arr): # TC O(n*2^n), SC O(n)
    res=[]
    def backtrack(idx,cur):
        if idx==len(arr):
            res.append(cur[:])
            return
        
        # include the current element
        cur.append(arr[idx])
        backtrack(idx+1,cur)
        cur.pop()

        # exclude the current element
        backtrack(idx+1,cur)
    backtrack(0,[])
    return res

print(generateSubsets(arr)) 